### DISCUSSION TITLE
ImGui Colors

### DISCUSSION THREAD

**Discussion by EricStancliff** [ocornut/imgui#3787] [Old/Archived discussions]
Hi!  I hope this is the right place for this question.  I have a two part.

1 - My colors (right side of image) seem to look different than the examples (left side) using the same style.  Is there any known reason why this would happen?  I've been using ImGui for a while with no major issues so I think my Vulkan renderer is handling it properly.

2 - Sometimes I get artifacts around multi line text inputs when I have it open and then open a few menus.  Has anyone ever seen this?

Thanks!!!
![ImGui](https://user-images.githubusercontent.com/25473977/106946450-b9c6df80-66ee-11eb-93bd-d5de98a2b85a.PNG)
![TextArtifacts](https://user-images.githubusercontent.com/25473977/106946464-bdf2fd00-66ee-11eb-9b70-68cb6b5ec896.PNG)

**Comment by ocornut** [comment]
> 1 - My colors (right side of image) seem to look different than the examples (left side) using the same style. Is there any known reason why this would happen? I've been using ImGui for a while with no major issues so I think my Vulkan renderer is handling it properly.

Try messing with the gamma (e.g. powf(1/2.2f) in your vertex shader).

> 2- Sometimes I get artifacts around multi line text inputs when I have it open and then open a few menus. Has anyone ever seen this?

Something is wrong about your render state: blending or z-testing.

**Reply by EricStancliff** [reply]
Thanks!  The z order was definitely the issue.  I was trying to set gl_fragDepth manually with a z value I jammed in the push constant prior to my engine supported ordered draws.  But now it does so I just moved it to that system and the artifacts are gone.

**Comment by EricStancliff** [comment]
Update - The text artifacts turned out to be a draw order issue which I've corrected.

**Comment by EricStancliff** [comment]
So applying a gamma correction seems to have pushed it further in the other direction, leading me to believe that maybe something in my pipeline is automatically applying a gamma correction?  Which seems implausible because it's Vulkan so since I didn't sweat it out for a week I'm sure it didn't "give" it to me.

![gamma](https://user-images.githubusercontent.com/25473977/106951730-c26ee400-66f5-11eb-93cf-4dfd726317af.PNG)
![vert](https://user-images.githubusercontent.com/25473977/106951785-d286c380-66f5-11eb-941a-89ed11368242.PNG)

**Comment by EricStancliff** [comment]
Ah ok, once I did what you said @ocornut, and UNDID the gamma correction (gamma = 1.f/2.2f), it now looks the same (see attached).  So the question becomes - Why do I have to do this when nobody else does?  What have I Done?  If I like the "corrected/wrong" look better, will that make things render incorrectly?

Thanks!
![UndoGamma](https://user-images.githubusercontent.com/25473977/106953293-d3205980-66f7-11eb-9459-62f611bf58b4.PNG)

**Reply by Tetane** [reply]
Hi, I think I have the same issue with the colors (but unlike you I don't have any artifacts).
I don't understand how you resolved the issue.
I render Dear ImGui in a separate renderPass and I haven't changed Dear ImGui pipeline or shaders.

EDIT:
I found a discussion around this issue : https://github.com/ocornut/imgui/issues/578
As you said, the issue comes from using VK_FORMAT_B8G8R8A8_SRGB.
I modified the vertex shader in "imgui/backends/vulkan" to implement a basic color correction (witch does not exactly give the same colors as the demo), generated the spv with "generate_spv.sh" and copied the text of the generated file inside "imgui_impl_vulkan.cpp".

**Comment by EricStancliff** [comment]
Ok I think it's because I'm using VK_FORMAT_B8G8R8A8_SRGB for my requested VkSurfaceFormatKHR for my swapchain, whereas the examples are using VK_FORMAT_R8G8B8A8_UNORM.  I feel better.  Thanks again for your help!
